bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/ReadOnlyUtils.java

```diff
@@ -351,8 +357,9 @@ private static BIntersectionType createAndSetImmutableIntersectionType(Module pk
      */
     public static Type getMutableType(BIntersectionType intersectionType) {
         for (Type type : intersectionType.getConstituentTypes()) {
-            if (intersectionType.getEffectiveType().getTag() == type.getTag()) {
-                return type;
+            Type referredType = TypeUtils.getReferredType(type);
```

Comment:
Shall we create an issue to add a test

Reply:
created https://github.com/ballerina-platform/ballerina-lang/issues/37112